pkg/api/message/v1/subscription.go

```diff
@@ -0,0 +1,157 @@
+		return nil, err
+	}
+	return dispatcher, nil
+}
+
+// Shutdown gracefully shuts down the dispatcher, unsubscribing from all topics.
+func (d *subscriptionDispatcher) Shutdown() {
+	_ = d.natsSub.Unsubscribe()
+	// the lock/unlock ensures that there is no in-process dispatching.
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.natsSub = nil
+	d.natsConn = nil
+	d.subscriptions = nil
+
+}
+
+// messageHandler processes incoming messages, dispatching them to the correct subscription.
+func (d *subscriptionDispatcher) messageHandler(msg *nats.Msg) {
+	var env proto.Envelope
+	err := pb.Unmarshal(msg.Data, &env)
+	if err != nil {
+		d.log.Info("unmarshaling envelope", zap.Error(err))
+		return
+	}
+
+	xmtpTopic := isValidSubscribeAllTopic(env.ContentTopic)
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	for subscription := range d.subscriptions {
+		if subscription.all && !xmtpTopic {
+			continue
+		}
+		if subscription.all || subscription.topics[env.ContentTopic] {
+			select {
+			case subscription.messagesCh <- &env:
+			default:
+				d.log.Info(fmt.Sprintf("Subscription message channel is full, is subscribeAll: %t, numTopics: %d", subscription.all, len(subscription.topics)))
+				// we got here since the message channel was full. This happens when the client cannot
+				// consume the data fast enough. In that case, we don't want to block further since it migth
+				// slow down other users. Instead, we're going to close the channel and let the
+				// consumer re-establish the connection if needed.
+				// close(subscription.messagesCh)
+				// delete(d.subscriptions, subscription)
+			}
+		}
+	}
+}
+
+// subscription represents a single subscription, including its message channel and topics.
+type subscription struct {
+	messagesCh chan *proto.Envelope    // Channel for receiving messages
+	topics     map[string]bool         // Map of topics to subscribe to
+	all        bool                    // Flag indicating subscription to all topics
+	dispatcher *subscriptionDispatcher // Parent dispatcher
+}
+
+// log2 calculates the base-2 logarithm of an integer using bitwise operations.
+// It returns the floor of the actual base-2 logarithm.
+func log2(n uint) (log2 uint) {
+	if n == 0 {
+		return 0
+	}
+
+	// Keep shifting n right until it becomes 0.
+	// The number of shifts needed is the floor of log2(n).
+	for n > 1 {
+		n >>= 1
+		log2++
+	}
+	return log2
+}
+
+// Subscribe creates a new subscription for the given topics.
+func (d *subscriptionDispatcher) Subscribe(topics map[string]bool) *subscription {
+	sub := &subscription{
+		dispatcher: d,
+	}
+
+	// Determine if subscribing to all topics or specific ones
+	for topic := range topics {
+		if natsWildcardTopic == topic {
+			sub.all = true
+			break
+		}
+	}
+	if !sub.all {
+		sub.topics = topics
+		// use a log2(length) as a backbuffer
+		backlogBufferSize := log2(uint(len(topics))) + 1
+		if backlogBufferSize < minBacklogBufferLength {
+			backlogBufferSize = minBacklogBufferLength
+		}
```

Comment:
I'm wondering if this was the source of the issue. `minBacklogBufferLength` doesn't seem to have been set

Reply:
In go constants get automatically set to the value in the constant above. In this case, the subscribe all limit

Reply:
I see! Okay I set it explicitly back to 1024